master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,609 @@
+        # 127 is the standard bash exit code for file-not-found.
+        return 127
+    except PermissionError:
+        # Unable to read or to execute the command.
+        shipper.ship_special(f"PermissionError executing {cmd}", metadata, emit_stdout_logs)
+        # 126 is the standard bash exit code for permission failure.
+        return 126
+    except Exception:
+        msg = f"unexpected failure executing {cmd}:\n" + traceback.format_exc()
+        shipper.ship_special(msg, metadata, emit_stdout_logs)
+        # 80 is the exit code we use to signal "ship_logs.py failed"
+        return 80
+
+    # Just for mypy.
+    assert isinstance(p.stdout, io.RawIOBase) and isinstance(p.stderr, io.RawIOBase)
+
+    try:
+        stdout = Collector(p.stdout, "stdout", emit_stdout_logs, metadata, logq, doneq)
+        stderr = Collector(p.stderr, "stderr", emit_stdout_logs, metadata, logq, doneq)
+        waiter = threading.Thread(target=lambda: doneq.put(("waiter", p.wait(), None)))
+
+        waiter.start()
+        waiter_started = True
+
+        # Set up signal forwarding.
+        def signal_passthru(signum: Any, frame: Any):
+            p.send_signal(signum)
+
+        for sig in [
+            signal.SIGINT,
+            signal.SIGTERM,
+            signal.SIGHUP,
+            signal.SIGUSR1,
+            signal.SIGUSR2,
+            signal.SIGWINCH,
+        ]:
+            signal.signal(sig, signal_passthru)
+
+        stdout.start()
+        stdout_started = True
+
+        stderr.start()
+        stderr_started = True
+
+        shipper.start()
+        shipper_started = True
+
+        # expect 4 messages on the doneq
+        first_who = None
+        first_error = None
+        exit_code = None
+        deadline = None  # type: Optional[float]
+        for _ in range(4):
+            # Wait for an event, possibly with a deadline (if the child process already exited).
+            try:
+                timeout = None if deadline is None else deadline - time.time()
+                if timeout is not None and timeout <= 0:
+                    raise queue.Empty()
+                who, what, error = doneq.get(timeout=timeout)
+            except queue.Empty:
+                # Deadline is done, just abandon the shipper.
+                shipper_timed_out = True
+                logging.error(
+                    f"waited {log_wait_time} seconds for shipper to finish after child exit; "
+                    "giving up now"
+                )
+                break
+
+            if who == "shipper":
+                # There's no point in collecting logs after the shipper is gone.
+                stdout.shipper_died = True
+                stderr.shipper_died = True
+
+            if who == "waiter":
+                # After the log shipper exits, the shipping code is on a deadline.
+                deadline = time.time() + log_wait_time
+                exit_code = what
+
+            if first_error is None and error is not None:
```

Comment:
> maybe because I don't know who is allowed to put errors into the doneq

`doneq` is where threads put their DoneMsg (in the new NamedTuple typing).  Every thread always puts into `doneq`.

For `first_error` and `first_who`... hm, let me work on that, I might be able to get the code to where we do just raise an error and clean up in the `finally`.

...

Ok that wasn't that hard.

Briefly I thought I could also just return `exit_code` as soon as the `who == "waiter"` donemsg arrives, but then I realized that risked dropping errors from the shipper as it ships the final logs from the process.

I wrote a comment explaining why such a simplification isn't safe, lmk if it is helpful.

Reply:
lol that's what you came up with, not what I came up with.  But thanks!